### PR TITLE
Plugin/module docs: parse return values, add collection names in them (version_added_collection), and format them nicely in ansible-doc

### PR DESCRIPTION
**Pull request by felixfontein**
##### SUMMARY
Right now, return value docs are treated as a string until the final stages (ansible-doc simply dumps them to the output, and plugin_formatter tries to parse the YAML). In particular, `version_added` in there are not tagged or untagged. (Which isn't really needed, since the versions should always refer to the current collection / ansible-base.)

This PR modifies the plugin docs processing so that the return value docs are also parsed in lib/ansible/parsing/plugin_docs.py and also tagged/untagged. Finally, it modifies ansible-doc so that it formats them nicer (using the same mechanism as for plugin/module options).

Combined with #69795 this gives a lot nicer representation of return values; for example (`ansible-doc acme_certificate_info`):
```
- account_uri
        ACME account URI, or None if account does not exist.

        returned: always
        type: str

- exists
        Whether the account exists.

        returned: always
        type: bool
```
instead of
```
exists:
  description: Whether the account exists.
  returned: always
  type: bool

account_uri:
  description: ACME account URI, or None if account does not exist.
  returned: always
  type: str
```
Also note the ordering - instead of just dumping the return values as described in the return value string, they will be sorted.

##### ISSUE TYPE
- Docs Pull Request
- Feature Pull Request

##### COMPONENT NAME
ansible-doc

### CODE REVIEW

**Comment by felixfontein** [comment]
@samdoran for modules, validate-modules does validate it to some regard (validation was improved for 2.9 and 2.10, and a lot of ignore.txt entries added IIRC). So it shouldn't be too bad.

Will add tests.

**Comment by samdoran** [comment]
@felixfontein I mainly just wanted to make sure this code is bit for flexible with the data it gets from the RETURN section so it doesn't explode when/if it encounters something odd.

**Comment by felixfontein** [comment]
Discussion from meeting (TBC): warn in case of unparsable YAML, or fail?

Note that this function is used in only three situations:
1. by ansible-doc
2. by ansible-console's help (it only cares for DOCUMENTATION and ignores RETURN)
3. by plugin_formatter (which will vanish soon)

File: lib/ansible/parsing/plugin_docs.py
Comment (bcoca) [inline lib/ansible/parsing/plugin_docs.py]: i would jsut put both in prev conditional `if theid in ('DOCUMENTATION, 'RETURN')`

File: test/integration/targets/ansible-doc/library/test_docs_returns.py
Comment (bcoca) [inline test/integration/targets/ansible-doc/library/test_docs_returns.py]: remove

**Review by bcoca — CHANGES_REQUESTED** [review]

File: test/integration/targets/ansible-doc/library/test_docs_returns.py
Comment (felixfontein) [reply test/integration/targets/ansible-doc/library/test_docs_returns.py]: done

File: lib/ansible/parsing/plugin_docs.py
Comment (felixfontein) [reply lib/ansible/parsing/plugin_docs.py]: As long as there is no consensus that bad YAML should result in an error, I'll keep it this way.

File: lib/ansible/parsing/plugin_docs.py
Comment (mattclay) [inline lib/ansible/parsing/plugin_docs.py:60]: In ansible-base we've required examples to be valid yaml for quite some time:

https://github.com/ansible/ansible/blob/840d3a9dd7d1233ab55aa9516fc9b44e4c865a81/test/lib/ansible_test/_data/sanity/yamllint/yamllinter.py#L171-L175

Comment (felixfontein) [reply lib/ansible/parsing/plugin_docs.py:60]: I guess the main concern are third-party plugins that weren't tested with `ansible-test`. OTOH this only affects `ansible-doc` and `ansible-console`'s help, so I think it is an acceptable risk.

**Comment by felixfontein** [comment]
I've adjusted the PR to always parse return docs as YAML, and fail if they cannot be parsed. A module with invalid return docs will return the following error in ansible-doc:
```
$ ansible-doc --playbook-dir ./ test_docs_returns_broken
[WARNING]: You are running the development version of Ansible. You should only run Ansible from "devel" if you are modifying the Ansible engine, or trying
out features under development. This is a rapidly changing source of code and can become unstable at any point.
ERROR! module test_docs_returns_broken missing documentation (or could not parse documentation): while parsing a block mapping
  in "<unicode string>", line 2, column 1
did not find expected key
  in "<unicode string>", line 4, column 4
```
Since all modules passing `ansible-test sanity` have valid return docs (see @mattclay's comment), I think it is acceptable to require this also for third-party modules. Especially since it only affects `ansible-doc` and `ansible-console`'s `help` command.

**Comment by acozine** [comment]
Discussed at the Docs Working Group meeting today, unanimously approved.

**Comment by felixfontein** [comment]
I've rebased the PR after #69926 was merged. This simplifies several parts, and also addresses the untagging placement in ansible-doc that @bcoca mentioned on IRC. (It's now in a better place, and also right now commented out - whether to re-enable it, and what exactly it should do, can be debated later, since it is not related to this PR.)

ready_for_review

File: lib/ansible/parsing/plugin_docs.py
Comment (bcoca) [reply lib/ansible/parsing/plugin_docs.py:60]: @mattclay  also, this is not just modules, but other plugins, you sometimes have an 'ini file' or other 'non yaml content' that goes in this section, so we should not restrict at docs code.

File: test/integration/targets/ansible-doc/collections/ansible_collections/testns/testcol/plugins/cache/notjsonfile.py
Comment (bcoca) [inline test/integration/targets/ansible-doc/collections/ansible_collections/testns/testcol/plugins/cache/notjsonfile.py]: this will create issues when loading the plugin, they should keep  FQCN name in docs

**Review by bcoca — CHANGES_REQUESTED** [review]

File: test/integration/targets/ansible-doc/collections/ansible_collections/testns/testcol/plugins/cache/notjsonfile.py
Comment (felixfontein) [reply test/integration/targets/ansible-doc/collections/ansible_collections/testns/testcol/plugins/cache/notjsonfile.py]: Hmm, then why are there no FQCNs in community.general for plugins?

File: lib/ansible/parsing/plugin_docs.py
Comment (felixfontein) [reply lib/ansible/parsing/plugin_docs.py:60]: @bcoca I *think* it checks all plugins, not just modules (despite the naming). Would have to check though...

Comment (mattclay) [reply lib/ansible/parsing/plugin_docs.py:60]: @felixfontein Yes, the `yamllint` sanity test does check the docs on all plugin types except `module_utils`, despite the name of the function being `check_modules`.

File: test/integration/targets/ansible-doc/collections/ansible_collections/testns/testcol/plugins/cache/notjsonfile.py
Comment (felixfontein) [reply test/integration/targets/ansible-doc/collections/ansible_collections/testns/testcol/plugins/cache/notjsonfile.py]: I've checked, and the plugin loader ignores it and doesn't pass this part of DOCUMENTATION on. The only place which seems to be using this is ansible-doc's man formatter, which uses it for the title.

Since it does the same for modules, and there FQCN is explicitly forbidden by ansible-test sanity, I think having the short name here is the better choice.

Comment (felixfontein) [reply test/integration/targets/ansible-doc/collections/ansible_collections/testns/testcol/plugins/cache/notjsonfile.py]: @bcoca if you insist, I can still change it. Just tell me.

Comment (mattclay) [reply test/integration/targets/ansible-doc/collections/ansible_collections/testns/testcol/plugins/cache/notjsonfile.py]: If the FQCN fails the 2.9 sanity tests then requiring it for 2.10 will create an issue for collections trying to support both versions.

File: test/integration/targets/ansible-doc/runme.sh
Comment (s-hertel) [inline test/integration/targets/ansible-doc/runme.sh:16]: I like having the additional --playbook-dir coverage since I think this was catching a legitimate bug.

oh, I overlooked the one right above it - nevermind me

**Review by s-hertel — COMMENTED** [review]
Looks good to me. It looks like the collection finder tries to load files that don't exist and requires absolute paths, so with this diff the collection plugin type tests pass again https://gist.github.com/s-hertel/1dc8415e3f5b24cd62c456eba1c0f83b.

File: test/integration/targets/ansible-doc/runme.sh
Comment (felixfontein) [reply test/integration/targets/ansible-doc/runme.sh:16]: This one actually didn't work because of unicode characters in Python 2. That's why I moved it (and the ones below, which I added in this PR) into the playbook.

Comment (s-hertel) [reply test/integration/targets/ansible-doc/runme.sh:16]: Ah, okay. That does seem like something we should fix (though I wouldn't hold this one up for it).

Comment (felixfontein) [reply test/integration/targets/ansible-doc/runme.sh:16]: Yes, and also the listing counting tests I had to comment out. :)

File: test/integration/targets/ansible-doc/collections/ansible_collections/testns/testcol/plugins/cache/notjsonfile.py
Comment (felixfontein) [reply test/integration/targets/ansible-doc/collections/ansible_collections/testns/testcol/plugins/cache/notjsonfile.py]: For modules, FQCN will fail both the 2.9 and 2.10 tests (if these aren't changed). For plugins, there is no test.

@abadger mentioned in #ansible-devel that the docs pipeline kind of depends (right now) on shortnames being used.

File: lib/ansible/parsing/plugin_docs.py
Comment (bcoca) [reply lib/ansible/parsing/plugin_docs.py:60]: @mattclay have we made exceptions for certain plugins? some are using ini files in examples ?

File: test/integration/targets/ansible-doc/collections/ansible_collections/testns/testcol/plugins/vars/noop_vars_plugin.py
Comment (felixfontein) [inline test/integration/targets/ansible-doc/collections/ansible_collections/testns/testcol/plugins/vars/noop_vars_plugin.py:5]: @bcoca this wasn't a FQCN before apparently.

File: lib/ansible/parsing/plugin_docs.py
Comment (mattclay) [reply lib/ansible/parsing/plugin_docs.py:60]: Not that I'm aware of.

Comment (bcoca) [reply lib/ansible/parsing/plugin_docs.py:60]:
```
EXAMPLES = r'''
    # define 2 hosts in command line
    # ansible -i '10.10.2.6, 10.10.2.4' -m ping all

    # DNS resolvable names
    # ansible -i 'host1.example.com, host2' -m user -a 'name=me state=absent' all

    # just use localhost
    # ansible-playbook -i 'localhost,' play.yml -c local
```
seems people have been 'creative' to avoid the restriction, we really should not make this a strict requirement

Comment (bcoca) [reply lib/ansible/parsing/plugin_docs.py:60]: specially since nothing will display if we treat it as yaml (currently its processed as string)

Comment (bcoca) [reply lib/ansible/parsing/plugin_docs.py:60]:
```
EXAMPLES = '''
  example1: |
      # example cfg file
      [web]
      host1
      host2 ansible_port=222 # defined inline, interpreted as an integer

      [web:vars]
```

**Review by bcoca — APPROVED** [review]

File: lib/ansible/parsing/plugin_docs.py
Comment (abadger) [reply lib/ansible/parsing/plugin_docs.py:60]: I know that examples is not valid structured yaml for a large number of plugins.  Perhaps yamllint decides it's a valid yaml string?

**Comment by felixfontein** [comment]
@samdoran @acozine @abadger @mattclay @bcoca @s-hertel thanks for reviewing and merging!

**Pull request merged**
